internal/broker/broker.go

```diff
@@ -476,10 +498,72 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *sessionInfo
 		return AuthNext, nil
 
 	case "password":
-		authInfo, err = b.loadAuthInfo(ctx, session, challenge)
+		var useOldEncryptedToken bool
+		exists, err := fileutils.FileExists(session.passwordPath)
 		if err != nil {
 			slog.Error(err.Error())
-			return AuthRetry, errorMessage{Message: "could not load cached info"}
+			return AuthDenied, errorMessage{Message: "could not check password file"}
+		}
+		if !exists {
+			// For backwards compatibility, we also check the old encrypted token path.
+			exists, err = fileutils.FileExists(session.oldEncryptedTokenPath)
+			if err != nil {
+				slog.Error(err.Error())
+				return AuthDenied, errorMessage{Message: "could not check old encrypted token path"}
+			}
+			if !exists {
+				return AuthDenied, errorMessage{Message: "password file does not exist"}
+			}
+			useOldEncryptedToken = true
+		}
+
+		if useOldEncryptedToken {
+			authInfo, err = tokenPkg.LoadOldEncryptedAuthInfo(session.oldEncryptedTokenPath, challenge)
+			if err != nil {
+				slog.Error(err.Error())
+				return AuthRetry, errorMessage{Message: "could not load old encrypted token"}
+			}
+		} else {
+			ok, err := password.CheckPassword(challenge, session.passwordPath)
+			if err != nil {
+				slog.Error(err.Error())
+				return AuthRetry, errorMessage{Message: "could not check password"}
+			}
+			if !ok {
+				return AuthRetry, errorMessage{Message: "invalid password"}
+			}
+
+			authInfo, err = tokenPkg.LoadAuthInfo(session.tokenPath)
+			if err != nil {
+				slog.Error(err.Error())
+				return AuthDenied, errorMessage{Message: "could not load cached info"}
+			}
+		}
+
```

Comment:
> The amount of nesting and conditionals here indicates that this should have its own function. handleIsAuthenticated does not care how the authInfo is loaded, it only cares about the object itself, so this logic could be moved into a loadAuthInfo function (like we had previously) and it seems like it doesn't even need to be attached to broker.

I'll try to extract something.

> Also, we are exposing way too much information on the error messages here. If you take a look at other modules (e.g. pam_unix) and SSH, we only get an "authentication failure" message to prevent giving too much information to a possible attacker (that's why we had only the "could not load auth info" message before).

I don't think the reason for a generic error message is to avoid giving too much information to an attacker. An attacker cannot do anything with the information that the password file does not exist or similar. The reason for a generic error message here is likely UX. The user doesn't know what "could not load auth info" means and what they can do about it. A good error message should explain what went wrong to the user in terms they can understand, and what they can do about it. A generic error message like "An unexpected error occurred. Please report this issue on <URL>" would be more useful to the user. 

On the other hand, a user facing such an error can't log in (at least not with the current user and authentication method), it might be hard for them to get the system logs and attach them when they report the problem. They might just report the error message they see instead. That's a reason to provide the information in the error message that *we* need to debug the issue.

So I think we should either go with a generic error message which the user can understand, or one which tells us exactly what went wrong, but "could not load auth info" is neither.

Reply:
> I'll try to extract something.

I extracted a `UseOldEncryptedToken` function to the `migration.go` file of the `token` package. I tried to also extract the whole `if useOldEncryptedToken { ... } else { ... }` block into a `CheckPasswordAndLoadAuthInfo` function, but it took too many arguments, and if we let `handleIsAuthenticated` return the same error message for any error returned by it, we would hide too much information.

Reply:
> I don't think the reason for a generic error message is to avoid giving too much information to an attacker

This is not a question of thinking. I told you what other modules do and they do not give you, the user trying to authenticate, any information as to why the authentication failed, only that it failed. We need to always keep in mind that we support SSH and so on, which means "anyone" could try and authenticate against authd on an exposed server (even worse, since we do go around one of the many layers regarding SSH safety measures [the user existence in the system]).

> On the other hand, a user facing such an error can't log in (at least not with the current user and authentication method), it might be hard for them to get the system logs and attach them when they report the problem. They might just report the error message they see instead. 

This is expected. We won't be able to get the logs unless a user, capable of authenticating, logs into the system and grabs them for us, same thing happens in other authenticators also (and we are already being quite gentle in what we print in the error messages).

> That's a reason to provide the information in the error message that we need to debug the issue.

We should not use this criteria to decide what we print to the _user_. That's why we log the original errors and print something else. If the user has a problem with authentication, they will likely talk to their IT admin before coming to us (or we will ask they to do so, as we'll need logs for it regardless of what error message we decide to print).

> So I think we should either go with a generic error message which the user can understand, or one which tells us exactly what went wrong, but "could not load auth info" is neither

That was not my point. I'm 100% down to use a better message if you have one.

Reply:
> This is not a question of thinking.

Of course it is. Other modules use a generic error message. Your interpretation is that they don't want to give too much information to an attacker. I gave you my interpretation. I try to avoid presenting my views and thoughts as facts, as is appropriate in a respectful conversation in my opinion.

> We need to always keep in mind that we support SSH and so on

Yes, we need to make sure that we don't leak any sensitive information or anything an attacker could use. I don't see anything that could be useful for an attacker in those error messages though. 

> That was not my point.

You refer to other modules which use a very generic error message like "authentication failure". My point is that we are not using such a generic error message, we already have "could not decode challenge", "could not get required response",  "could not authenticate user remotely",  "could not get ID token",  "could not fetch user info",  "could not load cached info",  "could not fetch user info",  "empty challenge",  "could not get required information", and "could not cache user info". I'm just adding to the already long list. I'm keeping the pattern. I'm not saying that we should keep doing that, and I explain why I think we should not. But if we decide to do it differently, let's please do so in a separate PR.

> I'm 100% down to use a better message if you have one.

 "An unexpected error occurred. Please report this issue." If the user is using a text interface (i.e. SSH or `sudo login` on the command-line, but not gdm), we can also include the URL which they can report issues at:  "An unexpected error occurred. Please report this issue on https://github.com/ubuntu/authd/issues".